File: dotnet/DiningPhilosophers/Waiter.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;

namespace DiningPhilosophers
{
    class Waiter
    {
        private RequestQueue _requestQueue = new RequestQueue();

        public async Task<bool> Request(Philosopher philosopher)
        {
            var completionSource = new TaskCompletionSource<bool>();
            EventManager.Subscribe(philosopher.Name + "RequestGranted", (name) => 
            {
                completionSource.SetResult(true);
            });
            _requestQueue.AddRequest(philosopher);
            await completionSource.Task.ConfigureAwait(false);
           
            return true;   
        }
    }
}

File: go/waiter.go
package main

import (
	"fmt"
)

type Waiter struct {
	requestQueue *RequestQueue
}

func NewWaiter() Waiter {
	q := NewRequestQueue()
	return Waiter{requestQueue: q}
}

func (waiter Waiter) Request(philosopher *Philosopher) chan bool {
	requestChan := make(chan bool)
	//defer close(requestChan)
	f := func(name string) {
		fmt.Printf("name: %s closing channel\n", name)
		requestChan <- true
		close(requestChan)
	}
	GetEventMangager().Subscribe(philosopher.Name+"RequestGranted", &f, true)
	waiter.requestQueue.AddRequest(philosopher)
	return requestChan
}
